contract-sdk/specs/oas20/types/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+}
+
+/// Token mintable information.
+#[derive(Debug, Default, Clone, cbor::Decode, cbor::Encode, PartialEq, Eq)]
+pub struct MintableInformation {
+    /// Caller address which is allowed to mint new tokens.
+    pub minter: Address,
+    /// Cap on the total supply of the token.
+    #[cbor(optional)]
+    pub cap: Option<u128>,
+}
+
+/// OAS20 token instantiation information.
+#[derive(Debug, Default, Clone, cbor::Decode, cbor::Encode, PartialEq, Eq)]
+pub struct TokenInstantiation {
+    /// Name of the token.
+    pub name: String,
+    /// Token symbol.
+    pub symbol: String,
+    /// Number of decimals.
+    pub decimals: u8,
+    /// Initial balances of the token.
+    #[cbor(optional, default, skip_serializing_if = "Vec::is_empty")]
+    pub initial_balances: Vec<InitialBalance>,
+    /// Information about minting in case the token is mintable.
+    #[cbor(optional)]
+    pub mintable: Option<MintableInformation>,
+}
+
+/// OAS20 token information.
+#[derive(Debug, Clone, cbor::Decode, cbor::Encode, PartialEq, Eq)]
+pub struct TokenInformation {
+    /// Name of the token.
+    pub name: String,
+    /// Token symbol.
+    pub symbol: String,
+    /// Number of decimals.
+    pub decimals: u8,
+    /// Total supply of the token.
+    pub total_supply: u128,
+    /// Information about minting in case the token is mintable.
+    #[cbor(optional)]
+    pub mintable: Option<MintableInformation>,
+}
+
+/// Struct passed to the "oas20_receive" method on the receiver contract when sending tokens via Oas20 Send.
+#[derive(Debug, Clone, cbor::Decode, cbor::Encode)]
+pub struct Receive {
+    pub sender: Address,
+    pub amount: u128,
+    pub data: cbor::Value,
+}
+
+/// All possible errors that can be returned by the OAS20 contract.
+///
+/// Each error is a triplet of (module, code, message) which allows it to be both easily
+/// human readable and also identifiable programmatically.
+#[derive(Debug, thiserror::Error, sdk::Error)]
+pub enum Error {
+    #[error("bad request")]
+    #[sdk_error(code = 1)]
+    BadRequest,
+
+    #[error("query failed")]
+    #[sdk_error(code = 2)]
+    QueryFailed,
+
+    #[error("balance overflow")]
+    #[sdk_error(code = 3)]
+    BalanceOverflow,
+
+    #[error("zero amount")]
+    #[sdk_error(code = 4)]
+    ZeroAmount,
+
+    #[error("insufficient funds")]
+    #[sdk_error(code = 5)]
+    InsufficientFunds,
+
+    #[error("action forbidden")]
+    #[sdk_error(code = 6)]
+    ActionForbidden,
+
+    #[error("mint over cap")]
+    #[sdk_error(code = 7)]
+    MintOverCap,
+}
+
+/// All possible events that can be returned by the OAS20 contract.
+#[derive(Debug, cbor::Encode, sdk::Event)]
+#[cbor(untagged)]
+pub enum Event {
+    #[sdk_event(code = 1)]
+    Oas20Instantiated { token_info: TokenInformation }, // TODO: include instance id?
+
+    #[sdk_event(code = 2)]
+    Oas20Transferer {
+        from: Address,
+        to: Address,
+        amount: u128,
+    }, // TODO: include token info, instance id?
```

Comment:
Hm, should we include instance ID as part of all contract events (e.g. where we do the mapping in the contracts module)? Code ID is already included as part of the module (e.g. the module for these events is `contracts.<code-id>`).

And for token info, any reason why this would need to be included in the event instead of just being queryable once you have the instance ID?

Reply:
Yeah if we have instance ID then the token info is not needed.  Will open an issue to include instance ID in all contract events, and remove the comments to include it here.